hash.ts

```diff
@@ -10,7 +10,148 @@ export type Description = Record<string, string | string[]> & {
 };
 
-const referenceLookups: Record<string, (e: Element) => Element[]> = {};
+/* currently only works if doName is set, GSSE support is missing */
+function findFCDAReferences(e: Element): Element[] {
+  const ldInst = e.getAttribute('ldInst');
+  const prefix = e.getAttribute('prefix');
+  const lnClass = e.getAttribute('lnClass');
+  const lnInst = e.getAttribute('lnInst');
+  const doName = e.getAttribute('doName');
+  const daName = e.getAttribute('daName');
+  const fc = e.getAttribute('fc');
+
+  if (!doName) {
+    return [];
+  }
+
+  const references = [] as Element[];
+
+  const server = e.closest('Server');
+  if (!server) {
+    return [];
+  }
+
+  const ln = server.querySelector(
+    `:scope>LDevice[inst="${ldInst}"]>LN0[prefix="${prefix}"][lnClass="${lnClass}"],
+     :scope>LDevice[inst="${ldInst}"]> LN[prefix="${prefix}"][lnClass="${lnClass}"][inst="${lnInst}"]`,
+  );
+
+  const lnType = ln?.getAttribute('lnType');
+
+  if (!ln || !lnType) {
+    return [];
+  }
+
+  const lNodeType = server
+    .closest('SCL')
+    ?.querySelector(`:scope>DataTypeTemplates>LNodeType[id="${lnType}"]`);
+
+  if (!lNodeType) {
+    return [];
+  }
+
+  let doNames = doName.replace(/\(\d+\)/g, '').split('.');
+  if (!doNames?.length) {
+    return [];
+  }
+
+  let doElement = lNodeType.querySelector(
+    `:scope>DO[name="${doNames.shift()}"]`,
+  );
+  let doTypeElement: Element | null = null;
+  while (doElement) {
+    const doType = doElement.getAttribute('type');
+    if (!doType) {
+      break;
+    }
+    const newDoTypeElement =
+      doElement
+        .closest('DataTypeTemplates')
+        ?.querySelector(`:scope>DOType[id="${doType}"]`) ?? null;
+    if (!newDoTypeElement) {
+      break;
+    }
+    doTypeElement = newDoTypeElement;
+    const sdoName = doNames.shift();
+    if (!sdoName) {
+      break;
+    }
+    doElement = doTypeElement.querySelector(`:scope>SDO[name="${sdoName}"]`);
+  }
+
+  if (!doElement) {
+    return [];
+  }
+
+  if (daName) {
+    const daNames = daName.replace(/\(\d+\)/g, '').split('.');
+    if (!daNames?.length) {
+      return [];
+    }
+
+    let daElement = doTypeElement?.querySelector(
+      `:scope>DA[name="${daNames.shift()}"]`,
+    );
+
+    while (daElement) {
+      const daType = daElement.getAttribute('type');
+      if (!daType) {
+        break;
+      }
+      const daTypeElement = daElement
+        .closest('DataTypeTemplates')
+        ?.querySelector(`:scope>DAType[id="${daType}"]`);
+      if (!daTypeElement) {
+        break;
+      }
+      const bdaName = daNames.shift();
+      if (!bdaName) {
+        break;
+      }
+      daElement = daTypeElement.querySelector(`:scope>BDA[name="${bdaName}"]`);
+    }
+    if (daElement) {
+      references.push(daElement);
+    } else {
+      return [];
+    }
+  } else if (fc) {
+    references.push(
+      ...Array.from(doElement.querySelectorAll(`:scope>DA[fc="${fc}"]`)),
+    );
+  }
```

Comment:
Perhaps a silly question:

> If I had an FCDA which only looked at at a DO, would the name of the DO somehow be included here, is this semantic? i.e. two boolean values with the same structure are the same even if the names are different?

This seems like some kind of duck typing vs. structural typing debate.
Somehow in my mind I feel the name of a data object is structural in a different way to, say, the name of an `IED` or `DataSet`.

Reply:
> This seems like some kind of duck typing vs. structural typing debate.
> Somehow in my mind I feel the name of a data object is structural in a different way to, say, the name of an IED or DataSet.
> 

I think I think this is because these are defined as part of the standard in Parts 7-3 and 7-4 and have a "well defined meaning". But up for discussion :wink:

Reply:
This is a very good point. If structural typing is desired, DO/DA names should be excepted from the base filter's exclude list.

Reply:
Perhaps we could ask @JakobVogelsang for a comment on this topic

Reply:
Just to be clear: Once #47 is resolved by @stee-re 's current work, whether we do structural typing or duck typing is up to the distributor to decide - or even up to the end user if they're willing to edit the base filter. Jakob's input would still be valuable for deciding which default base filter we want to deliver out of the box.